core/src/main/java/org/springframework/security/oauth2/server/authorization/PkceCodeChallengeMethodNames.java

```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.server.authorization;
+
+// https://tools.ietf.org/html/rfc7636#section-6.2.2
+// PKCE Code Challenge Method Registry > Initial Registry Contents
+// Method names are case-sensistive
+public interface PkceCodeChallengeMethodNames {
```

Comment:
Remove and add these as `private static` constants in `OAuth2AuthorizationCodeAuthenticationProvider`

Reply:
I don't know what's the best place to put those. Not sure this should go in OAuth2AuthorizationCodeAuthenticationProvider: they are used in multiple places.
- The CODE_VERIFIER constant is used by `OAuth2TokenEndpointFilter.AuthorizationCodeAuthenticationConverter` ; as it is a required parameter for public clients.
- The CODE_CHALLENGE and CODE_CHALLENGE_METHOD should also be checked in the Authorization endpoint, as per [RFC 7636](https://tools.ietf.org/html/rfc7636#section-4.4.1). For public clients, these parameters are required in the authorization request. Also, checking whether the server supports the transformation (for S256) should be done in the authorization endpoint. Relevant paragraph from the spec:

```
4.  Protocol
4.4.  Server Returns the Code
4.4.1.  Error Response
If the server requires Proof Key for Code Exchange (PKCE) by OAuth
public clients and the client does not send the "code_challenge" in
the request, the authorization endpoint MUST return the authorization
error response with the "error" value set to "invalid_request".  The
"error_description" or the response of "error_uri" SHOULD explain the
nature of error, e.g., code challenge required.
If the server supporting PKCE does not support the requested
transformation, the authorization endpoint MUST return the
authorization error response with "error" value set to
"invalid_request".  The "error_description" or the response of
"error_uri" SHOULD explain the nature of error, e.g., transform
algorithm not supported.
```